conn.go

```diff
@@ -356,17 +357,47 @@ func (c *Conn) Transport() tpt.Transport {
 	return c.config.transport
 }
 
+// Limit message size until we have a better
+// packetizing strategy.
+const dcWrapperBufSize = math.MaxUint16
+
 // dcWrapper wraps datachannel.ReadWriteCloser to form a net.Conn
 type dcWrapper struct {
 	channel datachannel.ReadWriteCloser
 	addr    net.Addr
+
+	buf      []byte
+	bufStart int
+	bufEnd   int
 }
 
 func (w *dcWrapper) Read(p []byte) (int, error) {
-	return w.channel.Read(p)
+	var err error
+
+	if w.bufEnd == 0 {
+		n := 0
+		n, err = w.channel.Read(w.buf)
```

Comment:
I must be missing something. We read from `w.buf` here but when do we ever write to it?

Reply:
This line means we read from the datachannel into the packet buffer.

The idea is to allow an upper layer to read the packet in chunks. This is allowed when reading from a stream but it causes you to lose data when reading on a packet based connection. I'm not sure if there can be better solutions without the upper layer taking into account that it's using a packet based connection. This also ties into a discussion of support for packet based transports in libp2p, e.g. UDP.

Reply:
Ah I was thrown off by the method name `Read`. But that line is actually "reading into" `w.buf` aka writing to it.